java-lib/src/main/java/com/yammer/metrics/core/WavefrontHistogram.java

```diff
@@ -91,18 +97,26 @@ public void update(int value) {
   }
 
   /**
-   * Helper to retrieve the current bin. Assumes proper synchronization.
+   * Helper to retrieve the current bin. Will be invoked per thread.
    */
   private MinuteBin getCurrent() {
-    long minMillis = minMillis();
-
-    if (bins.isEmpty() || bins.getLast().minMillis != minMillis) {
-      bins.add(new MinuteBin(minMillis));
-      if (bins.size() > MAX_BINS) {
-        bins.removeFirst();
+    long currMinMillis = minMillis();
+
+    long key = Thread.currentThread().getId();
+    LinkedList<MinuteBin> bins = perThreadHistogramBins.getOrDefault(key, new LinkedList<>());
```

Comment:
unfortunately you shouldn't do this since every get on an existing key would create a new linkedlist that would be discarded right away